### ISSUE TITLE
Simple envs for sanity checking

### ISSUE THREAD

**Issue by cathywu** [cathywu/rllab#9]
- No state env: there is no state, and the goal is to minimize the action vector.
- Multiaction point env: there is 1 agent, the overall action is the sum over k actions.
- Multiagent point env with collisions: same multiagent point env as before but terminates when two point masses get sufficiently close.

[Experiment snapshot](https://github.com/cathywu/rllab/blob/43f7576328b7cd6874d4db962dc352e8afa4e525/examples/cluster_multiagent_point_comparison.py) from commit 43f7576.

```
python3 examples/cluster_multiagent_point_comparison.py
```

**Comment by cathywu** [comment]
## Preliminary results  [exp cluster-multiagent-v10]

**Summary**: the action-dependent baseline does not improve/regress on training (the training curves look identical across all experiments so far), even though in the case of NoStateEnv with k=6, its explained variance is much larger.

NoStateEnv, with k=6, average return:
![2017-04-26-nostateenv-k6](https://cloud.githubusercontent.com/assets/329117/25449414/8a583d98-2a70-11e7-9bfa-9778129723e1.png)

NoStateEnv, with k=50, average return:
![2017-04-26-nostateenv-k50](https://cloud.githubusercontent.com/assets/329117/25449422/8d7d2132-2a70-11e7-801f-423365b73614.png)

MultiactionPointEnv, with k=6, average return:
![2017-04-26-multiactionpointenv-k6](https://cloud.githubusercontent.com/assets/329117/25449425/91289cbc-2a70-11e7-911c-8432d9c375b6.png)

NoStateEnv, with k=6, explained variance:
![2017-04-26-nostateenv-k6-explainedvariance](https://cloud.githubusercontent.com/assets/329117/25449433/93cfb2ca-2a70-11e7-857c-5dda34f48d82.png)
Furthermore, as a bonus, this somewhat confirms that I didn't accidentally just run with the same (non action dependent baseline) for all the experiments.

**Comment by cathywu** [comment]
New env: [OneStepNoStateEnv](https://github.com/cathywu/rllab/blob/efb7e17ffe5b0c078f8e0660b68d2f66bb0c12de/rllab/envs/one_step_no_state_env.py) (commit efb7e17)

**OneStepNoStateEnv**, with k=6, average return (the two curves are exactly overlapping):
![2017-04-26-onestepnostateenv-k6](https://cloud.githubusercontent.com/assets/329117/25457581/a82df554-2a8b-11e7-8fc6-48f2213badfc.png)

**OneStepNoStateEnv**, with k=6, explained variance (**huhhh?**):
![2017-04-26-onestepnostateenv-k6-explainedvariance](https://cloud.githubusercontent.com/assets/329117/25457653/db4cee90-2a8b-11e7-8f66-ea88e648c563.png)

**Comment by cathywu** [comment]
## Possible reasons action-dependent (AD) baseline not effective
- Envs are too easy.
- Envs are too low dimensional. [try k=200]
- We're taking so many samples per batch (on such simple problems), such that variance is sufficiently reduced enough to not need a AD baseline? [try batch sizes of 100, 500, 1000, 5000 -- **NOTE: these are currently running**]

**Comment by cathywu** [comment]
OneStepNoStateEnv, with k=6, **batch size=100**, average return (the two curves are exactly overlapping):
![2017-04-26-onestepnostateenv-k6-batch100](https://cloud.githubusercontent.com/assets/329117/25459229/30776af2-2a92-11e7-9996-ef36a28a0621.png)

OneStepNoStateEnv, with k=6, **batch size=100**, explained variance (huhhhhhh):
![2017-04-26-onestepnostateenv-k6-batch100-explainedvariance](https://cloud.githubusercontent.com/assets/329117/25459228/306c3330-2a92-11e7-88f4-6682a388ddd4.png)

The high explained variance (values of 1) of the GaussianMLPBaseline results from low variance among the return values, after a good policy has been learned, e.g.:
```
>> np.var(returns)
8.584473133996777e-09
```
The baseline is then overfitting to the only value it sees.

**Comment by cathywu** [comment]
### Summary: ZeroBaseline (with whitening) does better!  [exp cluster-multiagent-v11]

OneStepNoStateEnv, with k=6:
![2017-04-28-zerobaselinerocksonestepenv](https://cloud.githubusercontent.com/assets/329117/25553363/1c696062-2c66-11e7-8d3d-075fbb0c24d5.png)

OneStepNoStateEnv, with k=50:
![2017-04-28-zerobaselinerocksonestepenv-k50](https://cloud.githubusercontent.com/assets/329117/25553375/904def66-2c66-11e7-8c69-5e17b83d454b.png)

OneStepNoStateEnv, with k=200:
![2017-04-28-zerobaselinerocksonestepenv-k200](https://cloud.githubusercontent.com/assets/329117/25553377/91db3cda-2c66-11e7-924f-f397a8446e31.png)

**Hypothesis**: poor fits of the NN baselines (because the linear feature baselines below seem to appear to match the performance of the whitened ZeroBaseline here).

**Recommendation**: Ignore the NN baseline results in these runs.

From Rocky: `It shouldn't be a fitting problem. The center adv option should undo everything the baseline does, since it recenters the advantage estimates.`

OneStepNoStateEnv, with k=200, holdout loss for the NN baseline (vf):
![2017-04-29-onestepnostateenv-k200-holdoutloss-vf](https://cloud.githubusercontent.com/assets/329117/25557979/25c16b7a-2cd2-11e7-96a4-6e88b33eadd8.png)

OneStepNoStateEnv, with k=200, holdout loss for the NN baseline (vf0):
![2017-04-29-onestepnostateenv-k200-holdoutloss-vf0](https://cloud.githubusercontent.com/assets/329117/25557978/25bf0948-2cd2-11e7-91be-fa8e25f86dde.png)

**Comment by cathywu** [comment]
### Summary [exp cluster-multiagent-v12]
- Experiment: OneStepNoStateEnv
- ZeroBaseline (without whitening) does worse as expected.
- In low dimensions, linear baselines basically match performance of the whitened ZeroBaseline (above).
- In higher dimensions, the action dependent linear baseline does indeed perform better than the linear baseline.

I'm now running experiments with high dimensions to see a more pronounced effect (`k=1000, 2000`).
**EDIT (2017-05-02)**: Added figures for `k=500, 1000, 2000`.

OneStepNoStateEnv, with k=6, no whitening (`center_adv=False`):
![2017-04-29-onestepnostateenv-nowhitening-k6](https://cloud.githubusercontent.com/assets/329117/25557813/79603b0c-2cce-11e7-816f-a23df3ef5bd8.png)
OneStepNoStateEnv, with k=50, no whitening (`center_adv=False`):
![2017-04-29-onestepnostateenv-nowhitening-k50](https://cloud.githubusercontent.com/assets/329117/25557815/796d940a-2cce-11e7-9b49-fea3ccd2353d.png)
OneStepNoStateEnv, with k=200, no whitening (`center_adv=False`) (**First positive result!**):
![2017-04-29-onestepnostateenv-nowhitening-k200](https://cloud.githubusercontent.com/assets/329117/25557814/796d0314-2cce-11e7-8c78-8f8030bcbf7a.png)
OneStepNoStateEnv, with k=500, no whitening (`center_adv=False`) (**Positive result!**):
![2017-05-02-onestepnostateenv-nowhitening-k500](https://cloud.githubusercontent.com/assets/329117/25639771/f5a75b58-2f41-11e7-8f9d-78a4a7a15a53.png)
OneStepNoStateEnv, with k=1000, no whitening (`center_adv=False`) (Less conclusive.):
![2017-05-19-onestepnostateenv-nowhitening-k1000](https://cloud.githubusercontent.com/assets/329117/26244097/3f57439a-3c43-11e7-8af9-a72007f02f11.png)
OneStepNoStateEnv, with k=2000, no whitening (`center_adv=False`) (Hitting scaling issues.):
![2017-05-02-onestepnostateenv-nowhitening-k2000](https://cloud.githubusercontent.com/assets/329117/25639770/f5a43928-2f41-11e7-9d34-b51de2e734d7.png)

**Comment by cathywu** [comment]
### Summary [exp cluster-multiagent-v13]
- Setting: Done condition for the MultiagentPointEnv and MultiactionPointEnv is reaching the origin.
- MultiactionPointEnv: linear baseline does better than action-dependent baseline in higher dimensions. (We would have expected the opposite. **Why might there be regression?**)
- MultiagentPointEnv: action dependent baseline does significantly better than linear baseline in higher dimensions. (**Second positive result!**)
- MultiagentPointEnv: Does the action dependent baseline improve sample efficiency? I.e. can we get away with smaller batch sizes? The last 3 figures show that this experiment is totally inconclusive on this problem. (_Recommendation_: run more trials of this experiment.) 

<hr/>

MultiactionPointEnv, k=6, no whitening, done=reach origin:
![2017-05-02-multiactionpointenv-doneisreachorigin-k6](https://cloud.githubusercontent.com/assets/329117/25638702/189dc3f8-2f3e-11e7-9a8d-ec699004fdbe.png)
MultiactionPointEnv, k=1000, no whitening, done=reach origin:
![2017-05-02-multiactionpointenv-doneisreachorigin-k1000](https://cloud.githubusercontent.com/assets/329117/25638704/1a073472-2f3e-11e7-84c1-6ddaf8621c91.png)

<hr/>

MultiagentPointEnv, k=6, no whitening, done=reach origin:
![2017-05-02-multiagentpointenv-doneisreachorigin-k6](https://cloud.githubusercontent.com/assets/329117/25639049/5c89bf6c-2f3f-11e7-9f7f-198a6e9a62f1.png)
MultiagentPointEnv, k=50, no whitening, done=reach origin:
![2017-05-02-multiagentpointenv-doneisreachorigin-k50](https://cloud.githubusercontent.com/assets/329117/25639048/5c870114-2f3f-11e7-97e2-f3a815719cfc.png)
MultiagentPointEnv, k=200, no whitening, done=reach origin:
![2017-05-02-multiagentpointenv-doneisreachorigin-k200](https://cloud.githubusercontent.com/assets/329117/25639047/5c86f728-2f3f-11e7-82ec-6a152e8fe191.png)
MultiagentPointEnv, k=500, no whitening, done=reach origin:
![2017-05-02-multiagentpointenv-doneisreachorigin-k500](https://cloud.githubusercontent.com/assets/329117/25639050/5c999482-2f3f-11e7-9f42-feb713d84a17.png)

<hr/>

MultiagentPointEnv, k=200, no whitening, done=reach origin, **batch size=100**:
![2017-05-02-multiagentpointenv-doneisreachorigin-k200-batch_size100](https://cloud.githubusercontent.com/assets/329117/25639394/95ae947e-2f40-11e7-9816-0610fd72cd61.png)
MultiagentPointEnv, k=200, no whitening, done=reach origin, **batch size=500**:
![2017-05-02-multiagentpointenv-doneisreachorigin-k200-batch_size500](https://cloud.githubusercontent.com/assets/329117/25639395/95cb20da-2f40-11e7-9742-2a048f159be1.png)
MultiagentPointEnv, k=200, no whitening, done=reach origin, **batch size=1000**:
![2017-05-02-multiagentpointenv-doneisreachorigin-k200-batch_size1000](https://cloud.githubusercontent.com/assets/329117/25639396/95ccc9b2-2f40-11e7-845e-b3a74365d89c.png)
